stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+            &message[..32]
+        } else {
+            &message
+        };
+        let Some(signer_signature_hash) =
+            Sha512Trunc256Sum::from_bytes(signer_signature_hash_bytes)
+        else {
+            debug!("Signer #{}: Received a signature result for a signature over a non-block. Nothing to broadcast.", self.signer_id);
+            return;
+        };
+
+        // TODO: proper garbage collection...This is currently our only cleanup of blocks
+        self.blocks.remove(&signer_signature_hash);
+
+        // This signature is no longer valid. Do not broadcast it.
+        if !signature.verify(aggregate_public_key, &message) {
+            warn!("Signer #{}: Received an invalid signature result across the block. Do not broadcast it.", self.signer_id);
+            // TODO: should we reinsert it and trigger a sign round across the block again?
+            return;
+        }
+
+        let block_submission = if message == signer_signature_hash.0.to_vec() {
+            // we agreed to sign the block hash. Return an approval message
+            BlockResponse::accepted(signer_signature_hash, signature.clone()).into()
+        } else {
+            // We signed a rejection message. Return a rejection message
+            BlockResponse::rejected(signer_signature_hash, signature.clone()).into()
+        };
+
+        // Submit signature result to miners to observe
+        debug!(
+            "Signer #{}: submit block response {block_submission:?}",
+            self.signer_id
+        );
+        if let Err(e) = self.stackerdb.send_message_with_retry(block_submission) {
+            warn!(
+                "Signer #{}: Failed to send block submission to stacker-db: {e:?}",
+                self.signer_id
+            );
+        }
+    }
+
+    /// Process a sign error from a signing round, broadcasting a rejection message to stackerdb accordingly
+    fn process_sign_error(&mut self, e: &SignError) {
+        warn!(
+            "Signer #{}: Received a signature error: {e:?}",
+            self.signer_id
+        );
+        match e {
+            SignError::NonceTimeout(_valid_signers, _malicious_signers) => {
+                //TODO: report these malicious signers
+                debug!(
+                    "Signer #{}: Received a nonce timeout error.",
+                    self.signer_id
+                );
+            }
+            SignError::InsufficientSigners(malicious_signers) => {
+                debug!(
+                    "Signer #{}: Received a insufficient signers error.",
+                    self.signer_id
+                );
+                let message = self.coordinator.get_message();
+                let block = read_next::<NakamotoBlock, _>(&mut &message[..]).ok().unwrap_or({
+                        // This is not a block so maybe its across its hash
+                        // This jankiness is because a coordinator could have signed a rejection we need to find the underlying block hash
```

Comment:
This "jankiness" here and elsewhere belongs in its own function ;)

Reply:
Haha will do.